New scheduling algorithm

fcremo opened the pull request:
## Summary

This branch contains orchestra new action scheduling algorithm.

## Problem

We want an action scheduling algorithm that:

- can handle alternatives, aka "the GCC/libc situation"
- can determine scheduling feasibility ahead of time
- allows to schedule actions in parallel
- handles inconsistent states
  - e.g. gcc is installed but libc is not

### The GCC/libc situation

Builds can now depend on installation of any build from another component. 
We (rev.ng) use/plan to use this feature to:

- allow installing different builds of a dependency
- managing the dependency chain for bootstrapping cross compilers

By design, rev.ng orchestra configuration describes the dependency chain for cross compilers with loops.
This is because when bootstrapping a compiler we want this dependency chain:
```
gcc@stage2 -> libc@default -> gcc@stage1 -> libc@headers
```

Installing `stage2` causes `stage1` to be uninstalled, and installing `libc@default` causes `libc@headers` to be uninstalled.
Because of this if we used the above static dependency chain every time `gcc@stage2` is needed we would need to reinstall everything from `libc@headers` upwards.

Two ways to solve this problem are:

- "cutting" the dependency chain if `gcc@stage1` and/or `libc/headers` are not needed
- specifying that `libc@default` needs "any gcc" and that `gcc@stage1` needs "any libc", so that if `gcc@stage2` and `libc@default` are installed, we can satisfy these dependencies by having `gcc@stage2 -> libc@default` and `libc@default -> gcc@stage2`

The latter solution does imply generating a cyclic dependency cycle, but this is done only if all its components are already installed (a.k.a recognizing there's nothing to do).

## Solution

We first build an initial dependency graph, collecting all dependencies for the "root" actions the user requested (e.g. install `somecomponent`).
This graph contains some special "AnyOf" nodes which represent alternatives. Only one of the successord of these nodes should be picked to satisfy the dependency.

The core step of the algorithm finds assignments for those nodes such that either no cycles are found, or all nodes in the cycles are satisfied.
The search is done such that already installed components and preferred builds (`component~preferredbuild`) are tried first.
The algorithm assigns choices from different strongly connected components independently. This optimization reduces the search space size from intractable to very manageable (in my testing execution time goes from "never" finishing to ~0.1s for resolving `ui/cold-revng` dependency graph).

Then a couple of trivial simplification steps are performed on the graph, and the actions are scheduled opportunistically in topological order, allowing parallel execution.

## Conclusion

I would like to have some feedback/third party testing on it before merging it to master.
If I forgot explaining something important let me know :)

aleclearmind commented:
* [ ] The first commit should be dropped.
* [ ] Is the second commit necessary? Let's not put `exit(...)` in every command handler, let's use return values, which should already be how it works. See 91fcb1971b4d4a4a1714570438bd5ea6fba4a035.
* [ ] This is the MR where we make the recursive hash a property of the component as opposed to the build, is that right? Can we write that in a commit message? It's rather important.

fcremo commented:
I rebased on current master, now there's just one commit.
In the process I fixed the command invocations so that every command handler now `return`s an error code that is bubbled up to `main` and used as the process exit code.
I also mentioned the changes to the hash computation in the commit message, good point.

aleclearmind commented:
I'm now testing this!
Side note, I had to fix the commit message. Make sure to follow the 50/72 convention.

aleclearmind commented:
Merged in 8b8c8f8, thanks!

aleclearmind commented on orchestra/actions/install.py:
```diff
@@ -125,14 +106,59 @@ def _prepare_tmproot(self):
+    def _build_and_install(self, args):
         logger.info("Executing install script")
-        env = dict(self.environment)
-        env["RUN_TESTS"] = "1" if (self.build.test and test) else "0"
```
This change broke a very important feature.
Please try to be much more careful when rebasing.

aleclearmind replied:
Fixed in e5354de

aleclearmind commented on orchestra/actions/install.py:
```diff
@@ -31,38 +39,19 @@ def _run(self, args):
+            source = "binary archives"
+        elif self.allow_build:
+            self._build_and_install(args)
+            if getattr(args, "create_binary_archives", False):
```
This is not OK, this need to become an argument

fcremo replied:
You're right, fixed in PR #9

aleclearmind commented on orchestra/actions/install.py:
```diff
@@ -31,38 +39,19 @@ def _run(self, args):
-            if args.create_binary_archives:
+        install_start_time = time.time()
+        if self.allow_binary_archive and self._binary_archive_exists():
+            self._install_from_binary_archive(args)
```
The patch is hard to read, but I think you're no longer calling `update_binary_archive_symlink` after installing from binary archives. It is very important.
You have to be extra careful when doing this kind of rebase.

fcremo replied:
I don't think you got the right line numbers for this comment, however, I do call `update_binary_archive_symlink` when a new binary archive is built. Is it necessary to call it regardless?
I was under the impression that its job is to update the link to the newly built binary archive

aleclearmind replied:
I'm not talking about *building* a new binary archive, I'm talking about installing. This is why the comment is positioned here.

The old code was:
```python
        if self.from_binary_archives and self._binary_archive_exists():
            ...
            self.update_binary_archive_symlink()
         elif not self.from_binary_archives or self.fallback_to_build: 
            ...
             if args.create_binary_archives: 
                 self._create_binary_archive() 
             self.update_binary_archive_symlink() 
         else: 
             raise OrchestraException("Binary archive not found!") 
```

While now you're only updating symlinks if the archive is being built from source.
This is a bit subtle and opportunistic but it's necessary for the CI, in order to update symlinks after `next-branch` is promoted to `branch`, which is something we want even if we're installing from binary archives.
This is rather important.

fcremo replied:
Ok I'll fix this right now, I didn't understand that the CI needed this behaviour

aleclearmind replied:
Yeah, it wasn't obvious, maybe add a comment about this to avoid such a situation in the future.

fcremo replied:
This commit should fix the issue https://github.com/revng/revng-orchestra/pull/9/files/e5f579a8d35264212095fd53515b96f9aadf928f..bdfa2922a4a21f2312fbaba4d2e7e2f94b58648b

aleclearmind commented on orchestra/actions/install.py:
```diff
@@ -125,14 +106,59 @@ def _prepare_tmproot(self):
-        env = dict(self.environment)
-        env["RUN_TESTS"] = "1" if (self.build.test and test) else "0"
-        run_script(self.script, quiet=quiet, environment=env)
+        run_script(self.script, quiet=args.quiet, environment=self.environment)
```
Do not pass the whole `args` on, you're coupling completely unrelated parts of the code

aleclearmind replied:
Is this being handled somewhere?

aleclearmind replied:
Can you update #9 to fix this too?

fcremo replied:
I believe this is fixed by #6, as `quiet` is entirely removed and `InstallAction` almost does not user `args` anymore (it was  removed from many other places as well)

aleclearmind commented:
I understand this PR had a long history of begin out-of-tree, but it broke in a subtle way multiple important things.
Please try to be extra careful when merging.

aleclearmind commented on orchestra/actions/install.py:
```diff
@@ -255,8 +281,8 @@ def _binary_archive_repo_name(self):
             binary_archive_repo_name = self.component.binary_archives
             if binary_archive_repo_name not in self.config.binary_archives_remotes.keys():
-                raise Exception(f"The {self.component.name} component wants to push to an unknown binary-archives repository ({binary_archive_repo_name})")
-            return binary_archive_repo_name
```
You dropped a `return`! Gesù!
Fixed in 5351b9cb7d3a222aaa52106146da079f2ee440b0.

aleclearmind commented on orchestra/cmds/clone.py:
```diff
@@ -22,10 +21,13 @@ def handle_clone(args):
-    return executor.run(build.component.clone, no_force=args.no_force)
+        return 1
+
+    executor = Executor(args, [build.clone])
```
`build.component.clone`

aleclearmind commented on orchestra/cmds/clone.py:
```diff
@@ -12,7 +12,6 @@ def install_subcommand(sub_argparser):
                                           parents=[execution_options],
                                           )
     cmd_parser.add_argument("component", help="Name of the component to clone")
-    cmd_parser.add_argument("--no-force", action="store_true", help="Don't force execution of the root action")
```
Why are we dropping this? Now if I try to clone an already cloned component, it will fail.

fcremo replied:
The new scheduling algorithm deprecated `--no-force`, it already did nothing.
The actions requested by the user are always performed, of course if you already cloned a component cloning it again will fail. 
I noticed that there's also another `--no-force` for the configure command which should also be removed.
If `--no-force` is needed I can reintroduce it with a slight modification to Executor
